### ISSUE TITLE
`relative_url_root` not respected

### ISSUE THREAD

**Issue by arianf** [rails/rails#42243]
The docs here say `config.action_controller.relative_url_root`, which contradicts the docs once you follow that link
<img width="616" alt="Screen Shot 2021-05-17 at 2 32 50 PM" src="https://user-images.githubusercontent.com/7397857/118559422-c2cb4200-b71c-11eb-9661-a855acfca44b.png">

This one says, `config.relative_url_root`
https://guides.rubyonrails.org/configuring.html#deploy-to-a-subdirectory-relative-url-root
<img width="677" alt="Screen Shot 2021-05-17 at 2 34 08 PM" src="https://user-images.githubusercontent.com/7397857/118559558-f1491d00-b71c-11eb-828f-19793232ac82.png">

So which doc is correct? hmmm is it `config.relative_url_root` or `config.action_controller.relative_url_root`?

But who cares, lets say you set both of these in the `config/environment/development.rb` :

```
  config.relative_url_root = '/sub'
  config.action_controller.relative_url_root = '/sub'
```

then you run puma without setting the `ENV['RAILS_RELATIVE_URL_ROOT']`...

```
$ puma
(byebug) MyApp::Application.config.relative_url_root
"/sub"

(byebug) MyApp::Application.config.action_controller.relative_url_root
"/sub"

(byebug) MyApp::Application.routes.url_helpers.projects_url
"http://localhost/projects"

(byebug) MyApp::Application.routes.url_helpers.projects_path
"/projects"
```

So you can see above it doesn't work... now if I add in `RAILS_RELATIVE_URL_ROOT="/sub" puma`

```
$ RAILS_RELATIVE_URL_ROOT="/sub" puma

(byebug) MyApp::Application.routes.url_helpers.projects_url
"http://localhost/sub/projects"

(byebug) MyApp::Application.routes.url_helpers.projects_path
"/sub/projects"
```

You can see you are required to put the value into ENV. The docs seem to be incorrect. See issue: https://github.com/rails/rails/issues/24393 and https://github.com/rails/rails/issues/22074

I think it likely has to do with this:

Railties only look at the ENV variable and it doesn't follow the docs at all and it doesn't bother to check the value at `config.relative_url_root` or `config.action_controller.relative_url_root` https://github.com/rails/rails/blob/5ddff2beb1a819d6b300fbef2731cba9c5f71db2/railties/lib/rails/application/configuration.rb#L51

**Comment by SkipKayhil** [comment]
Hey @arianf, I tried reproducing the issue with this script, but both of the assertions pass. Could you provide an example repo/script to reproduce the issue?

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", '6.1.3.2'
end

require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  config.hosts << "example.org"
  secrets.secret_key_base = "secret_key_base"

  ###
  config.relative_url_root = '/sub'
  config.action_controller.relative_url_root = '/sub'
  ###

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    default_url_options host: "localhost:3000"

    resources :posts
    get "/path" => "test#path"
    get "/url" => "test#url"
  end
end

class TestController < ActionController::Base
  include Rails.application.routes.url_helpers

  def path
    render plain: TestApp.routes.url_helpers.posts_path
  end

  def url
    render plain: TestApp.routes.url_helpers.posts_url
  end
end

require "minitest/autorun"
require "rack/test"

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def test_posts_path
    get "/path"
    assert_equal last_response.body, "/sub/posts"
  end

  def test_posts_url
    get "/url"
    assert_equal last_response.body, "http://localhost:3000/sub/posts"
  end

  private
    def app
      Rails.application
    end
end
```

**Comment by arianf** [comment]
@SkipKayhil here is your error case

https://github.com/arianf/relative-root

`$ puma`

<img width="1177" alt="Screen Shot 2021-05-17 at 4 40 54 PM" src="https://user-images.githubusercontent.com/7397857/118569358-a7693280-b72e-11eb-9e09-4379a09de615.png">

`$ RAILS_RELATIVE_URL_ROOT="/sub" puma`
<img width="1177" alt="Screen Shot 2021-05-17 at 4 41 12 PM" src="https://user-images.githubusercontent.com/7397857/118569381-b223c780-b72e-11eb-942f-9b80dce2c614.png">

**Comment by pixeltrix** [comment]
Firstly, regarding your point about `config.action_controller.relative_url_root` - that's what the configuration used to be called in earlier versions of Rails but has been moved to the general application config because it affects a number of areas in Rails - Action Mailer, Assets as well as Action Controller. It seems we missed one place to rename it in the guides and I apologise for that.

Secondly, the line that sets `RAILS_RELATIVE_URL_ROOT` in `configuration.rb` is merely setting the default value - it can still be overridden by using `config.relative_url_root = '/path'` in your application. However what's not clear in the documentation is that to affect url generation the config needs to be set early enough so that the config passed the `ActionDispatch::Router::RouteSet.new` has it set - setting it too late has no effect. Taking your example app, if you move the setting into the `config/application.rb` file then the routes will be generated with the prefix (this explains why @SkipKayhil's tests were passing - the config was set before the routes were drawn). I'm unsure whether this is a regression or it's always needed to be this early and it'll require further investigation to track this down.

Thirdly, I just want to address the following in `config.ru`:

``` ruby
# without this nothing works... lol
map RelativeRoot::Application.config.relative_url_root || '/' do
  run Rails.application
end
```

There's a very good reason why you need that there and it relates to the CGI environment variables [`SCRIPT_NAME`][1] and [`PATH_INFO`][2]. The former is what you'd consider the relative url root and Rails needs Rack to that set correctly when the app receives `call` with the request environment hash - if it's not set then you might as well just wrap all your routes in a `scope("/sub") { ... }` block. Using `map` creates a `Rack::URLMap` instance which [you can see from the code][3] sets `SCRIPT_NAME` and `PATH_INFO` to the desired values. When a Rails application has a mounted Engine (or other Rack app), those values then get adjusted so that you can easily mount them at different URLs without altering the Engine. The `SCRIPT_NAME` value takes precedence over the relative url config when there is a request context like in a controller but in other contexts like an Action Mailer view or an Active Job task there is no `SCRIPT_NAME` so there the relative url config is used. I hope this helps you understand the complex interactions at play here - it's not a simple matter of just prefixing `config.relative_url_root` onto every generated url.

Finally, can I please ask that you refrain from using language like ‘lied’ - this suggests intent on our part to mislead you for some unspecified reason and whilst we're all fallible and miss errors or oversights in documentation we're certainly not doing it on purpose.

[1]: https://datatracker.ietf.org/doc/html/rfc3875#section-4.1.13
[2]: https://datatracker.ietf.org/doc/html/rfc3875#section-4.1.5
[3]: https://github.com/rack/rack/blob/master/lib/rack/urlmap.rb

**Comment by arianf** [comment]
Thanks for your detailed response! 

> Secondly, the line that sets `RAILS_RELATIVE_URL_ROOT` in `configuration.rb` is merely setting the default value - it can still be overridden by using `config.relative_url_root = '/path'` in your application. However what's not clear in the documentation is that to affect url generation the config needs to be set early enough so that the config passed the `ActionDispatch::Router::RouteSet.new` has it set - setting it too late has no effect. Taking your example app, if you move the setting into the `config/application.rb` file then the routes will be generated with the prefix (this explains why @SkipKayhil's tests were passing - the config was set before the routes were drawn). I'm unsure whether this is a regression or it's always needed to be this early and it'll require further investigation to track this down.

Mmm, this makes a lot of sense! Thank you for breaking this down for me! The docs don't really show where to set `config.relative_url_root`, I assume it should be in the `environments/<env>.rb` file, since production subdirectory might be different than development. Was that a valid assumption to make?


> I hope this helps you understand the complex interactions at play here - it's not a simple matter of just prefixing config.relative_url_root onto every generated url.

This definitely helps, appreciate you going into detail to explain this! I guess my only suggestion would be to update the Rails docs to include this explanation, and maybe include this solution for when this would be needed.  https://guides.rubyonrails.org/configuring.html#deploy-to-a-subdirectory-relative-url-root

> this suggests intent on our part to mislead you for some unspecified reason and whilst we're all fallible and miss errors or oversights in documentation we're certainly not doing it on purpose.

My apologizes, didn't mean it in a negative way or to imply that this was done out of malice, only that following the current docs doesn't work. But I've updated my original issue to be more civilized. Sorry about that!

**Comment by arianf** [comment]
Also since Puma is the default web server in Rails now, maybe the docs should have a puma example where this example is shown

```ruby
map RelativeRoot::Application.config.relative_url_root || '/' do
  run Rails.application
end
```

**Comment by arianf** [comment]
> I'm unsure whether this is a regression or it's always needed to be this early and it'll require further investigation to track this down.

Just tested rails `3.2.22.5` and it works without an ENV var:

https://github.com/arianf/relative-root-3

<img width="956" alt="Screen Shot 2021-05-17 at 10 45 08 PM" src="https://user-images.githubusercontent.com/7397857/118597445-88d05f00-b761-11eb-8934-2f392f331f85.png">

**Comment by arianf** [comment]
Actually this is only an issue when calling the helper method directly instead of invoking it with just the helper method

```console
$ puma
Puma starting in single mode...
* Puma version: 5.3.1 (ruby 2.5.1-p57) ("Sweetnighter")
*  Min threads: 5
*  Max threads: 5
*  Environment: development
*          PID: 900
* Listening on http://0.0.0.0:3000
Use Ctrl-C to stop
```
<img width="956" alt="Screen Shot 2021-05-17 at 10 58 45 PM" src="https://user-images.githubusercontent.com/7397857/118598764-748d6180-b763-11eb-8d03-fd7ae0a1b4bc.png">

**Comment by arianf** [comment]
Hmm, the reason why I call the method directly, instead of by the short hand name, is because I need to refer to a parent route path/url from within an engine. Interesting

**Comment by f-mer** [comment]
I tracked down this issue further and would like to share my observations.
My expectation is that calling `root_path` within controllers and views would return a string starting with `"/sub"` if `ENV['RAILS_RELATIVE_URL_ROOT'] = "/sub"` is set. Instead `root_path` returns `"/"`.

The confusion stems from [`rails/actionpack/lib/action_controller/metal/url_for.rb`](https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/actionpack/lib/action_controller/metal/url_for.rb#L46-L49) where `script_name` is set to `"/"`.
If now `root_path` or any other route helper is called within the context of a controller the `url_options[:script_name]` superseeds the configured `relative_url_root`.
This can be seen [here](https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/actionpack/lib/action_dispatch/routing/route_set.rb#L787).

This behaviour can be circumvented by removing `:script_name`.

```rb
class ApplicationController < ActionController::Base
  def url_options
    super.except(:script_name)
  end
end
```

This test case demonstrates the issue:

```rb
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "rails", "6.1.0"
end

require "rack/test"
require "action_controller/railtie"

require "minitest/autorun"

class BugTest < Minitest::Test
  def test_returns_success
    relative_url_root = "/foo"
    config = ::ActionDispatch::Routing::RouteSet::Config.new(relative_url_root, false)
    route_set = ::ActionDispatch::Routing::RouteSet.new(config)
    route_set.draw do
      root to: "test#index"
    end
    helpers = Class.new do
      include route_set.url_helpers

      def url_options
        super.merge!(script_name: "/")
      end
    end

    assert helpers.new.root_path.start_with?(relative_url_root), "Expected #{helpers.new.root_path} to start with #{relative_url_root}"
  end
end
```

**Comment by zzak** [comment]
@f-mer Thanks for your investigation. I'm taking a stab at putting this down in the docs over at #42303, would you mind taking a look and giving any feedback? :bow:

**Comment by cfurrow** [comment]
Thanks @f-mer, as a temporary work around, that worked.

I set `config.relative_url_root = '/myapp'` in `config/environments/production.rb` (as its only needed in production for my use case), and I put that config at the very top, right under `Rails.application.configure do`

```ruby
# config/environments/production.rb
require "active_support/core_ext/integer/time"

Rails.application.configure do
  config.relative_url_root = '/myapp'
  #...snip
```
Then I added your changes to `ApplicationController#url_options`, and that seemed to do the trick for me:

```ruby
class ApplicationController < ActionController::Base
  def url_options
    super.except(:script_name)
  end
end
```

My prod server is running Puma, so this all works now. 

An additional note, my assets were working just fine before adding the change to `url_options`, and I think that was noted elsewhere that how assets, controllers/routes, Rack-apps use `config.relative_url_root` may vary a little. I also may have misunderstood the comments, as they spanned multiple issue numbers.

**Comment by techgique** [comment]
Hi all. I've been trying to get a Rails 5.2.x app running via Puma and Passenger Standalone reverse-proxied via Apache at a sub-RUI and having trouble getting links to work. It works fine at a sub-URI via Passenger+Apache deployment, but these seem to be missing something and I might have found a hint but don't know where to go from there.

**Background**
`RAILS_RELATIVE_URL_ROOT="/3000" rails s` for Puma and Passenger standalone with `"envvars": { "RAILS_RELATIVE_URL_ROOT": "/3000" },` both fail to prefix link URLs with `/3000` as expected. The only URLs it does help with are static asset URLs so at least the CSS and background images load. Adding `config.relative_url_root = "/3000"` to the Rails environment or application config files doesn't fix `link_to` links either. So I wound up searching the web and ended up here.

@pixeltrix 's addition in `config.ru` didn't help. Replace `RelativeRoot` in his example with your app module name from `config/application.rb` in case anyone doesn't remember and wants to try it out. Oddly with that change in place, I had to send a request to Apache for `https://(host.domain.tld)/3000/3000` to get the app's home page to load but then all the links are still broken.

@f-mer 's addition of
```ruby
  def url_options
    super.except(:script_name)
  end
```
to `application_controller.rb` as well as the env var and config file configuration gets things the closest to fully working. All the `link_to` links correctly have the sub-URI prepended with this in place. The only thing that doesn't work is the app is using Devise for auth and its redirects after sign in and sign out are broken and result in redirections to URLs like `https://(host.domain.tld)/30003000`. I suspect this is a little more the fault with Devise but I didn't give up yet.

**New Hint?**
I looked around for what could be differing between the Puma/Passenger Standalone deployment vs the Passenger+Apache deployment. I entered in a bad URL request intentionally to get the Rails error page to show and noticed that looking at the "Toggle env dump" contents had a difference.

Passenger+Apache:
`ORIGINAL_SCRIPT_NAME: "/nebraska_authors"`
Puma or Passenger Standalone:
`ORIGINAL_SCRIPT_NAME: ""`

Perhaps Rails is failing to set this somewhere and this could help with others' woes where the documentation's guidance is unsuccessful for deploying at a sub-URI :crossed_fingers: Maybe someone here has an idea on whether this is part of the problem the `url_options` hack works on. :v:
